### PR TITLE
Support exposing the Audience of a Broker

### LINKED ISSUE #3520: [control-plane]: Support exposing the Audience of a Broker
As the [Eventing OIDC feature track](https://docs.google.com/document/d/1e7UgNTkL0Br5Da09Rg2ieVmhKJo4VXuBj-mHT9NCujY/edit#heading=h.glanb1rq6iu5) describes, Addressables (and thus Brokers) need to announce their audience in `.status.address.audience`.

In particular the EKB broker controller should:

* when the feature flag (see https://github.com/knative/eventing/issues/7174) is disabled:
  * no change in behavior
* when the feature flag (see https://github.com/knative/eventing/issues/7174) is enabled:
  * Set the Brokers `.status.address.audience` accordingly.

Additional Information:
* In eventing-core we have a helper function to create the audience string so the follow the same pattern: https://github.com/knative/eventing/blob/1e84b2e5328017d34e8c0c8e4ddd84da67381526/pkg/auth/audience.go#L28-L32 and its usage in https://github.com/knative/eventing/blob/1e84b2e5328017d34e8c0c8e4ddd84da67381526/pkg/reconciler/broker/broker.go#L242-L249

### PR DESCRIPTION
**Pull request by gunishmatta**
Fixes #3520 

## Proposed Changes

- Support exposing the Audience of a Broker based on flag **IsOIDCAuthentication**

### CODE REVIEW

**Comment by pierDipi** [comment]
Thanks for the PR @gunishmatta, please check linter errors https://github.com/knative-extensions/eventing-kafka-broker/actions/runs/7521296763/job/20481144990?pr=3600

**Comment by gunishmatta** [comment]
@pierDipi  I have fixed the formatter, can you please approve the workflow run?

**Comment by Leo6Leo** [comment]
@gunishmatta Thanks for the PR! As we can see from the error log of the failing tests, we can see there problems with your code. Hope this information give you some hint!
```
containerID: containerd://783e4cd89babfca38b1dec1b2fafc1e65d30ccbe2ca0c444a3f64891c2d6bee3
      image: sha256:47ff6e6f1028017bd4e6fefb9b5046f630f6913d16a7750dbd91da68804e6bb7
      imageID: gcr.io/knative-boskos-80/eventing-kafka-broker-e2e-img/7587/kafka-controller-409c76cb00d7d7c46941c7e4c366e4a6@sha256:cfb93cd819ad68831e609a727e848822f10413ed458f53e58ef473c5e19b215d
      lastState:
        terminated:
          containerID: containerd://783e4cd89babfca38b1dec1b2fafc1e65d30ccbe2ca0c444a3f64891c2d6bee3
          exitCode: 2
          finishedAt: "2024-01-15T12:39:41Z"
          message: "cileKind(0xc00007d4a0, {0x2591bc8, 0xc001805620}, 0xc000fbaa80)\n\tknative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/broker/broker.go:271
            +0x2836\nknative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/broker.(*Reconciler).ReconcileKind.func1()\n\tknative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/broker/broker.go:90
            +0x25\nk8s.io/client-go/util/retry.OnError.func1()\n\tk8s.io/client-go@v0.28.5/util/retry/util.go:51
            +0x30\nk8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtection(0x2202727?)\n\tk8s.io/apimachinery@v0.28.5/pkg/util/wait/wait.go:145
            +0x3e\nk8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x989680,
            0x4014000000000000, 0x3fb999999999999a, 0x4, 0x0}, 0xc000fbab98?)\n\tk8s.io/apimachinery@v0.28.5/pkg/util/wait/backoff.go:461
            +0x5a\nk8s.io/client-go/util/retry.OnError({0x989680, 0x4014000000000000,
            0x3fb999999999999a, 0x4, 0x0}, 0x2202727?, 0x5?)\n\tk8s.io/client-go@v0.28.5/util/retry/util.go:50
            +0xa5\nk8s.io/client-go/util/retry.RetryOnConflict(...)\n\tk8s.io/client-go@v0.28.5/util/retry/util.go:104\nknative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/broker.(*Reconciler).ReconcileKind(0x6?,
            {0x2591bc8?, 0xc001805620?}, 0xc000fbab98?)\n\tknative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/broker/broker.go:89
            +0xb5\nknative.dev/eventing/pkg/client/injection/reconciler/eventing/v1/broker.(*reconcilerImpl).Reconcile(0xc0002654a0,
            {0x2591bc8, 0xc0018055c0}, {0xc0016cbf60, 0xd})\n\tknative.dev/eventing@v0.39.1-0.20240111175334-3cbddd640f53/pkg/client/injection/reconciler/eventing/v1/broker/reconciler.go:251
            +0x5f9\nknative.dev/pkg/controller.(*Impl).processNextWorkItem(0xc0005e8a20)\n\tknative.dev/pkg@v0.0.0-20240111013350-bc230ae58d14/controller/controller.go:542
            +0x4ad\nknative.dev/pkg/controller.(*Impl).RunContext.func3()\n\tknative.dev/pkg@v0.0.0-20240111013350-bc230ae58d14/controller/controller.go:491
            +0x59\ncreated by knative.dev/pkg/controller.(*Impl).RunContext in goroutine
            365\n\tknative.dev/pkg@v0.0.0-20240111013350-bc230ae58d14/controller/controller.go:489
            +0x349\n"
          reason: Error
          startedAt: "2024-01-15T12:39:09Z"
```

**Comment by creydr** [comment]
Hello @gunishmatta,
can we help you somehow on this?

**Comment by gunishmatta** [comment]
@creydr  sorry was busy with work during the week, will debug it today

**Comment by gunishmatta** [comment]
@creydr  can you please approve workflow, have added tests and fixed issues locally.

**Comment by gunishmatta** [comment]
Hi Team @pierDipi  @creydr , Please give a final review to the PR and provide the feedback.

Thanks a lot

Looking forward to contribute more to knative community.

**Review by creydr — DISMISSED** [review]
Hello @gunishmatta,
thanks for your PR. I have 2 comments so far:

1. we should provision the audiences in the `.status.addresses` too
2. Currently it does not set/clear the audience, when the feature flag config map gets updated again. You can check on https://github.com/knative/eventing/pull/7126 how the broker was updated to reconcile broker resources, when the feature flags config map was updated.

**Review by Leo6Leo — COMMENTED** [review]
Just to provide you with more information to help you better understand the approach that @creydr shared!

https://github.com/knative/eventing/blob/bb5313df8dd5295573cbc4a08206af741edeeef0/pkg/reconciler/broker/broker.go#L246

Also please run gofmt and goimports to fix the linter issues. Don't forget to run it everytime before you push your changes!

**Comment by gunishmatta** [comment]
Thanks for the feedback, I will be working on the feedback, you can expect an update within a few days.

**Comment by creydr** [comment]
Hey @gunishmatta,
can I help you somehow on this? 
I am working on exposing the audience from a KafkaChannel in #3622. Maybe this helps you a bit.

File: control-plane/pkg/reconciler/broker/broker.go [hunk 1]
```diff
@@ -282,6 +283,18 @@ func (r *Reconciler) reconcileKind(ctx context.Context, broker *eventing.Broker)
 
 	broker.Status.Address = addressableStatus.Address
 	broker.Status.Addresses = addressableStatus.Addresses
+
+	if feature.FromContext(ctx).IsOIDCAuthentication() && broker.Status.Address != nil {
+		audience := auth.GetAudience(eventing.SchemeGroupVersion.WithKind("Broker"), broker.ObjectMeta)
+		logging.FromContext(ctx).Debugw("Setting the brokers audience", zap.String("audience", audience))
+		broker.Status.Address.Audience = &audience
```
Comment (creydr) [inline control-plane/pkg/reconciler/broker/broker.go:290]: can you also set the audience in `broker.Status.Addresses[*].Audience`? (and clear/nil them too if OIDC is disabled)

**Comment by gunishmatta** [comment]
HI @creydr very sorry, was occupied with work commitments last week, I will try to get out time and update the PR before this weekend.

Thanks for the support.

**Comment by Leo6Leo** [comment]
@gunishmatta Hey Gunish, are you being blocked by any issue? anything we could help you with on this PR?

**Comment by creydr** [comment]
@gunishmatta Since it seemed you were occupied with other work and there was not much left to do, I added the missing bits from my [comment](https://github.com/knative-extensions/eventing-kafka-broker/pull/3600#discussion_r1469864348) in 6d9b33d839855e1653c839bc809f7a43a5acc575 (and run gofmt & goimports 5f7bd8bee9537a496a3d2a11f7db7a11daf5d8b6). I hope that's ok.

@Leo6Leo looks OK from my side. Can you recheck (as I committed to this PR now too :smile: )

/assign @Leo6Leo

**Comment by creydr** [comment]
@Leo6Leo can you PTAL

**Review by creydr — APPROVED** [review]

**Pull request merged**